KinesisStreamProvisioner is overwriting headerMode property

tomek82 opened the pull request:
When provisioning a stream consumer or producer the passed in binder properties get overwritten with HeaderMode.none.
The issue arises when using multiple destinations (i.e. a default function router for two streams) and not using multiplexing where in such case only the first stream will be able to parse/write the headers.

tomek82 committed:
make sure the embedded headers property does not get ovewritten

Signed-off-by: Tomek Bielecki <tomasz.bielecki@gmail.com>

tomek82 committed:
don't overwrite the consumer/producer properties

Signed-off-by: Tomek Bielecki <tomasz.bielecki@gmail.com>

artembilan commented:
Did you investigate why it had been done like that ? Sorry, not on the laptop to dig into the code .

tomek82 commented:
This was introduced in https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/issues/185 but I'm not sure why.

artembilan commented on spring-cloud-stream-binder-kinesis/src/main/java/org/springframework/cloud/stream/binder/kinesis/provisioning/KinesisStreamProvisioner.java:
```diff
@@ -86,11 +86,6 @@ public ProducerDestination provisionProducerDestination(String name,
-		properties.setHeaderMode(HeaderMode.none);
-
-		if (properties.getHeaderMode() == null) {
-			properties.setHeaderMode(HeaderMode.embeddedHeaders);
```
Yes, I cannot recall right now why I did that reset, but it definitely has to be `HeaderMode.embeddedHeaders` by default for Kinesis since there is "headers" abstraction over there, but we still need to carry out headers information somehow.

So, any reason to remove this default setting?

tomek82 replied:
Since those would previously default to HeaderMode.none I was afraid it would be a breaking change.

Defaulting to HeaderMode.embeddedHeaders breaks tests in org.springframework.cloud.stream.binder.kinesis.observation.KinesisBinderObservationTests#observationIsPropagatedFromSupplierToConsumer

and

org.springframework.cloud.stream.binder.AbstractBinderTests: testSendAndReceiveNoOriginalContentType and testSendAndReceive

since the encoded payloads come with a header and tests are validating against raw payload values before decoding.

artembilan commented on spring-cloud-stream-binder-kinesis/src/main/java/org/springframework/cloud/stream/binder/kinesis/provisioning/KinesisStreamProvisioner.java:
```diff
@@ -103,7 +98,6 @@ public ConsumerDestination provisionConsumerDestination(String name, String grou
 		KinesisConsumerProperties kinesisConsumerProperties = properties.getExtension();
 		kinesisConsumerProperties.setEmbedHeaders(
 				properties.getHeaderMode() == null || HeaderMode.embeddedHeaders.equals(properties.getHeaderMode()));
-		properties.setHeaderMode(HeaderMode.none);
```
I think we need to do similar setting for `HeaderMode.embeddedHeaders` on consumer side as well.
Since again: the "embedded" mode is default for Kinesis.

artembilan reviewed (changes requested):
Please, find some simple review on line.

And sorry that my unreasonable change caused some problems on your side.

Thank you!

artembilan reviewed (changes requested):
Well, I remember now why I did that.
You see, there is an embedded headers functionality in the Spring Cloud Stream Core.
The logic is like this in the `AbstractMessageChannelBinder`:
```
		if (HeaderMode.embeddedHeaders.equals(properties.getHeaderMode())) {
			bindingTarget.addInterceptor(0, this.embeddedHeadersChannelInterceptor);
		}
```
According to the problem described in the mentioned https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/issues/185, we cannot rely on that for tracing headers. Therefore, the logic of embedded headers has been moved to the respective Kinesis channel adapter implementation. 

I would love to ignore that `HeaderMode` from Spring Cloud Stream Core, but that's not what is possible from this project perspective.

Maybe for now you can workaround in your project that `spring.cloud.stream.default.<producer|consumer>.headerMode=<value>` to the `spring.cloud.stream.kinesis.default.<producer|consumer>.embedHeaders=true|false`?

I think we can raise a GH issue against Spring Cloud Stream Core  to consider to deprecate `embeddedHeaders` option in favor of custom one in the respective Binder implementation.

artembilan commented:
> The issue arises when using multiple destinations (i.e. a default function router for two streams)

Would you mind demonstrating your use-case to understand problem better?

tomek82 commented:
My setup is like so:

`
spring.cloud.function.definition=functionRouter
spring.cloud.function.routing-expression=headers['test_header']
spring.cloud.stream.function.routing.enabled=true
spring.cloud.stream.bindings.functionRouter-in-0.destination=test-stream1,test-stream2
spring.cloud.stream.bindings.functionRouter-in-0.consumer.header-mode=embeddedHeaders
spring.cloud.stream.kinesis.bindings.functionRouter-in-0.consumer.embed-headers=true
`

When using multiplexing:
`
spring.cloud.stream.bindings.functionRouter-in-0.consumer.multiplex=true
`
this works fine. However this means the consumer is shared (therefore same thread) between streams which is not ideal if the throughput varies a lot between streams.

When multiplexing is not enabled the test-stream2 will not be parsing the payloads and decoding the headers because at the time the consumer gets created the header mode is reset to none by the provisioning code for test-stream1.

Adding 
`
spring.cloud.stream.default.consumer.headerMode=embeddedHeaders
spring.cloud.stream.kinesis.default.consumer.embedHeaders=true
`
doesn't change anything here.

artembilan commented:
I see. I'm not familiar with that `functionRouter`, but looks like we cannot mitigate with existing logic, unless you can stay with `multiplex=true` for time being.

As a fix I suggest to rely only `kinesis.bindings.functionRouter-in-0.consumer.embed-headers=true` and fully ignore value of the `consumer.headerMode` and override it to `none` if `embed-headers == true` .
Make `KinesisConsumerProperties.embedHeaders` and `KinesisProducerProperties.embedHeaders` as `true` by default.
And in that `KinesisStreamProvisioner` have a logic like:
```
		KinesisProducerProperties kinesisProducerProperties = properties.getExtension();
		if (kinesisProducerProperties.isEmbedHeaders()) {
			properties.setHeaderMode(HeaderMode.none);
		}
```
instead of what we have so far.
Even if we override that `consumer|producer.headerMode`, we just don't deal with it and with the fact that `embed-headers == true`  we do embedding on the Kinesis Binder level, not in upstream Spring Cloud Stream Core logic.

tomek82 committed:
Changing the way header embedding settings are processed while
provisioning streams.

Signed-off-by: Tomek Bielecki <Tomek.Bielecki@aus.com>

tomek82 commented:
Sorry I was not being clear, `functionRouter` is the [RoutingFunction from Spring Cloud Function project](https://docs.spring.io/spring-cloud-function/docs/current/reference/html/spring-cloud-function.html#_function_routing_and_filtering).

I've implemented your suggestions and also added handling of `useNativeDecoding` and `useNativeEncoding` and modified tests to work with new defaults.

artembilan commented:
@tomek82 ,

thank you for contribution; looking forward for more!
